src/_P039_Thermocouple.ino

```diff
@@ -484,12 +736,865 @@ int Plugin_039_convert_two_complement(uint32_t value, int nr_bits) {
+    @param refResistor The value of the matching reference resistor, usually
+    430 or 4300
+    @returns Temperature in C
+*/
+/**************************************************************************/
+float Plugin_039_convert_to_temperature(uint32_t l_rawvalue, float RTDnominal, float refResistor)
+{
+
+  #define RTD_A 3.9083e-3
+  #define RTD_B -5.775e-7
+
+  float Z1, Z2, Z3, Z4, Rt, temp;
+
+  Rt = l_rawvalue;
+  Rt /= 32768;
+  Rt *= refResistor;
+
+  Z1 = -RTD_A;
+  Z2 = RTD_A * RTD_A - (4 * RTD_B);
+  Z3 = (4 * RTD_B) / RTDnominal;
+  Z4 = 2 * RTD_B;
+
+  temp = Z2 + (Z3 * Rt);
+  temp = (sqrt(temp) + Z1) / Z4;
+
+  if (temp >= 0)
+    return temp;
+
+  Rt /= RTDnominal;
+  Rt *= 100; // normalize to 100 ohm
+
+  float rpoly = Rt;
+
+  temp = -242.02;
+  temp += 2.2228 * rpoly;
+  rpoly *= Rt; // square
+  temp += 2.5859e-3 * rpoly;
+  rpoly *= Rt; // ^3
+  temp -= 4.8260e-6 * rpoly;
+  rpoly *= Rt; // ^4
+  temp -= 2.8183e-8 * rpoly;
+  rpoly *= Rt; // ^5
+  temp += 1.5243e-10 * rpoly;
+
+  return temp;
+}
+
+uint16_t getNomResistor(uint8_t l_RType)
+{
+  uint16_t l_returnValue = 100;
+
+  switch (l_RType)
+  {
+    case MAX31865_PT100:
+                          l_returnValue = 100;
+                          break;
+    case MAX31865_PT1000:
+                          l_returnValue = 1000;
+                          break;
+    default:
+                          l_returnValue = 100;
+                          break;
+  }
+  return (l_returnValue);
+}
+
+float readLM7x(struct EventStruct *event)
+{
+
+  float temperature = 0;
+  uint16_t device_id = 0;
+  uint16_t rawValue = 0;
+
+  uint8_t CS_pin_no = get_SPI_CS_Pin(event);
+
+  // operate LM7x devices in polling mode, assuming conversion is ready with every call of this read function ( >=210ms call cycle)
+  // this allows usage of multiples generations of LM7x devices, that doe not provde conversion ready information in temperature register
+
+  rawValue = readLM7xRegisters(CS_pin_no, P039_RTD_LM_TYPE, P039_RTD_LM_SHTDWN, &device_id);
+
+  temperature = convertLM7xTemp(rawValue, P039_RTD_LM_TYPE);
+
+  # ifndef BUILD_NO_DEBUG
+
+    if (loglevelActiveFor(LOG_LEVEL_DEBUG))
+    {
+      String log;
+      log.reserve(66);
+      log = F("P039 : LM7x : readLM7x : ");
+      log += F(" rawValue: ");
+      log += String(rawValue, DEC);
+      log += F(" device_id: 0x");
+      log += String(device_id, HEX);
+      log += F(" temperature: ");
+      log += String(temperature, DEC);
+      addLog(LOG_LEVEL_DEBUG, log);
+    }
+
+  # endif // ifndef BUILD_NO_DEBUG
+
+  return (temperature);
+}
+
+float convertLM7xTemp(uint16_t l_rawValue, uint16_t l_LM7xsubtype)
+{
+  float l_returnValue = 0;
+  float l_lsbvalue = 0;
+  uint8_t l_noBits = 0;
+  int l_intTemperature = 0;
+
+  switch (l_LM7xsubtype)
+  {
+    case LM7x_SD70:
+                        l_rawValue >>= 5;
+                        l_lsbvalue = 0.25;
+                        l_noBits = 11;
+                        break;
```

Comment:
Float values should be post-fixed with a `f`, and usually, also the `.0` is added before that if no decimals are specified.
```
float l_lsbvalue = 0.0f;
...
l_lsbvalue = 0.25f;
```

Reply:
Solved - i searched the code for constant value suffixes where missing.